Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,504 @@
+    }
+
+    /// This enables the connection pool to store and retreive `ConnectionProvider`s
+    /// by ensuring thread safety using the `lock`.
+    private subscript(key: Key) -> ConnectionProvider? {
+        get {
+            return self.lock.withLock {
+                _connectionProviders[key]
+            }
+        }
+        set {
+            self.lock.withLock {
+                _connectionProviders[key] = newValue
+            }
+        }
+    }
+
+    /// Determines what action should be taken regarding the the `ConnectionProvider`
+    ///
+    /// This method ensures there is no race condition if called concurently from multiple threads
+    private func getAction(for key: Key, eventLoop: EventLoop) -> ProviderGetAction {
+        return self.lock.withLock {
+            if let provider = self._connectionProviders[key] {
+                return .useExisting(provider)
+            } else {
+                let promise = eventLoop.makePromise(of: ConnectionProvider.self)
+                self._connectionProviders[key] = .future(promise.futureResult)
+                promise.futureResult.whenComplete { result in
+                    switch result {
+                    case .success(let provider):
+                        self[key] = provider
+                    case .failure:
+                        self[key] = nil
+                    }
+                }
+                return .make(promise)
+            }
+        }
+    }
+
+    /// This method asks the pool for a connection usable by the specified `request`, respecting the specified options.
+    ///
+    /// - parameter request: The request that needs a `Connection`
+    /// - parameter preference: The `EventLoopPreference` the connection pool will respect to lease a new connection
+    /// - parameter deadline: The connection timeout
+    /// - Returns: A connection  corresponding to the specified parameters
+    ///
+    /// When the pool is asked for a new connection, it creates a `Key` from the url associated to the `request`. This key
+    /// is used to determine if there already exists an associated `ConnectionProvider` in `connectionProviders`
+    /// if it does, the connection provider then takes care of leasing a new connection. If a connection provider doesn't exist, it is created.
+    func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, deadline: NIODeadline?) -> EventLoopFuture<Connection> {
+        let key = Key(request: request)
+        // FIXME: Is this correct to use self.loopGroup?
+        switch self.getAction(for: key, eventLoop: self.loopGroup.next()) {
+        case .useExisting(let provider):
+            return provider.getConnection(preference: preference)
+
+        case .make(completing: let providerPromise):
+            // If no connection provider exists for the given key, create a new one
+            let selectedEventLoop = providerPromise.futureResult.eventLoop
```

Comment:
Rather than using the loop from here, if we throw this away as I've suggested above we could simply use the `preference` to give us an event loop for this initial connection attempt. That improves efficiency in the cases where the user does have a preference but we don't have a connection.